### ISSUE TITLE
Recommended way to serialize <template> elements?

### ISSUE THREAD

**Issue by vinsonchuong** [jsdom/jsdom#3783]
This isn't an issue report or feature request. It's a question about how best to make a specific usecase happen.

I find JSDOM super convenient for quickly throwing together prototypes for things involving parsing and building HTML documents. Currently, I'm exploring [Declarative Shadow DOM](https://web.dev/articles/declarative-shadow-dom) and trying to build something that server-side renders a bunch of Web Components to an HTML string. This involves outputting a bunch of `<template shadowrootmode="open">` tags that contain the rendered output of the Web Components.

Now, I know that `<template>` tags aren't "actually rendered" to the DOM. So, `jsdom.serialize()` just outputs empty `<template>` tags. I've been able to work around this by manually setting the `innerHTML` of the `<template>` elements that I create. That seems to cause `jsdom.serialize()` to output what I want.

My question is, is this a sane way to achieve my usecase? Or, is there some better strategy I should be using instead? It seems a bit lucky that this works for me at all.

Thanks!

**Comment by alexyuly** [comment]
I would like to do the same thing. You're not the only one, at least :)

**Comment by WorldMaker** [comment]
I stepped into this use case as well. It also doesn't look like `jsdom.serialize()` renders `<slot>` tags at all, as well. It does complicate my SSR plans, I think.

**Comment by asamuzaK** [comment]
Is this what you want to achieve?

``` javascript
const dom = new JSDOM('', {
  runScripts: 'dangerously'
});
const document = dom.window.document;

// create template element
const template = document.createElement('template');
template.setAttribute('shadowrootmode', 'open');

// create template content nodes
const div = document.createElement('div');
const span = document.createElement('span');
span.textContent = 'Bar';
div.append(document.createTextNode('Foo'), span);
const div2 = document.createElement('div');
div2.textContent = 'Baz';

// append content nodes to `template.content`, not to `template`
template.content.append(div, div2);
document.body.appendChild(template);

// outputs the serialized DOM string with template content
console.log(dom.serialize());

// `document.documentElement.outerHTML` also outputs the serialized DOM string
console.log(dom.serialize() === document.documentElement.outerHTML);
```

**Comment by WorldMaker** [comment]
@asamuzaK Yes, in my case I want to make sure that the contents of the template are included in a `dom.serialize()`. (I don't think I'm worried about `outerHTML` in my usecases at the moment, just `dom.serialize()`.)

Is appending to `template.content` instead of `template` a current work-around? Is that the preferred behavior, even? 

(It does look like a cleaner fix than assigning a clone to `template.innerHTML`.)

Is the `shadowrootmode` attribute required in every case case of using `template.content` to get `dom.serialize()` to work?

I suppose that *I* still don't entirely "get" the Shadow DOM and `template.content` versus `template.childNodes` nor understand all their various interactions and browsing circles on MDN haven't yet illuminated what I'm missing.

**Comment by WorldMaker** [comment]
Verified that appending to `template.content` instead of `template` did the trick in my current test cases. I did not set any `shadowrootmode` attribute. Still curious if this is the "proper" behavior.

**Comment by asamuzaK** [comment]
> (I don't think I'm worried about `outerHTML` in my usecases at the moment, just `dom.serialize()`.)

`dom.serialize()` is JSDOM only.
`document.documentElement.outerHTML` is a web standard, it can be used in browsers, happy-dom, deno-dom, etc.

> Is appending to `template.content` instead of `template` a current work-around? Is that the preferred behavior, even?

`template.content` is a DocumentFragment and subtrees are held there.
See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template#template_document_fragment

> (It does look like a cleaner fix than assigning a clone to `template.innerHTML`.)

If you already have stringified HTML for the subtree, using innerHTML is fine.

> Is the `shadowrootmode` attribute required in every case case of using `template.content` to get `dom.serialize()` to work?

No, the question was about how to serialize `<template shadowrootmode="open">`, so I added `shadowrootmode` attribute to the sample above.

**Comment by asamuzaK** [comment]
``` html
<div id="host">
  <template id="template" shadowrootmode="open">
    <style>
      div {
        color: green;
      }
    </style>
    <div></div>
    <slot></slot>
  </template>
</div>
```
is kinda equivalent to
``` html
<div id="host">
  <template id="template">
    <style>
      div {
        color: green;
      }
    </style>
    <div></div>
    <slot></slot>
  </template>
</div>
<script>
  const template = document.getElementById('template');
  const host = document.getElementById('host');
  const shadowRoot = host.attachShadow({ mode: 'open' });
  shadowRoot.appendChild(template.content.cloneNode(true));
</script>
```

**Comment by TechQuery** [comment]
https://github.com/EasyWebApp/declarative-shadow-dom-polyfill can help you guys~

**Comment by domenic** [comment]
This is indeed proper behavior per the HTML Standard. Closing.
